extensions/amp-viewer-integration/0.1/examples/amp-viewer-host.js

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {APP, Messaging, MessageType} from '../messaging';
+import {listen} from '../../../../src/event-helper';
+import {dev} from '../../../../src/log';
+
+/**
+ * @fileoverview This class is a de-facto implementation of MessagePort
+ * from Channel Messaging API:
+ * https://developer.mozilla.org/en-US/docs/Web/API/Channel_Messaging_API
+ */
+class WindowPortEmulator {
+  /**
+   * @param {!Window} win
+   * @param {!Window} ampdoc
+   * @param {string} origin
+   */
+  constructor(win, ampdoc, origin) {
+    /** @const {!Window} */
+    this.win = win;
+    /** @const {!Window} */
+    this.ampdoc_ = ampdoc;
+    /** @private {string} */
+    this.origin_ = origin;
+  }
+
+  /**
+   * @param {string} eventType
+   * @param {function(!Event):undefined} handler
+   */
+  addEventListener(eventType, handler) {
+    listen(this.win, 'message', e => {
+      console.log('%%% got a message from the amp doc!', e.data);
+      if (e.origin == this.origin_ && e.data.app == APP) {
+        console.log('Viewer is about to handle message!');
+        handler(e);
+      }
+    }.bind(this));
+  }
+
+  /**
+   * @param {Object} data
+   */
+  postMessage(data) {
+    console.log('&&& viewer posting message to amp doc', data);
+    this.ampdoc_./*OK*/postMessage(data, this.origin_);
+  }
+  start() {
+  }
+}
+
+
+
+const CHANNEL_OPEN_MSG = 'channelOpen';
+
+/**
+ * @fileoverview This is an example of how the viewer host can be implemented
+ * for communication with the AMP docs.
+ */
+export class AmpViewerHost {
+
+  /**
+   * @param {!Window} win
+   * @param {!HTMLIFrameElement} ampIframe
+   * @param {string} viewerOrigin
+   * @param {boolean} startPolling
+   */
+  constructor(win, ampIframe, viewerOrigin, startPolling, opt_id) {
+    /** @const {!Window} */
+    this.win = win;
+    /** @const {!HTMLIFrameElement} */
+    this.ampIframe_ = ampIframe;
+
+    /** @const {string} */
+    this.id = opt_id;
```

Comment:
What is this ID and why needed? Also, it's not declared above and opt_ suggests optionality, but it's not defined as `=`. Maybe this is something we can read from `iframe.id`?